### PR TITLE
ChainDB.StateMachine: testing wiping the VolatileDB

### LINKED ISSUE #1628: Chain DB: simulate VolatileDB corruption
#1625 fixes #1624, but doesn't test it, because it's quite an edge case to test it: the VolatileDB has to be empty and the block we're adding must be an EBB with the same block as the block at the current tip.

Suggested approach: add a `WipeVolDB` command to the ChainDB q-s-m tests that closes the ChainDB, wipes the entire VolatileDB, and reopens the ChainDB. This could trigger #1624.

Doing partial truncation of the VolatileDB is much more complicated, as it will depend on implementation details of the VolatileDB.

### PR DESCRIPTION
**Pull request by mrBliss**
Fixes #1628.

Please review commit per commit.

### CODE REVIEW

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs [hunk 1]
```diff
@@ -440,10 +481,64 @@ instance UpdateLedger TestBlock where
...
-  -- Use defaults
+  type TipInfo TestBlock = IsEBB
+  getTipInfo = thIsEBB . unTestHeader
 
 instance ValidateEnvelope TestBlock where
-  -- Use defaults
+  -- A chain can start with an EBB with block number 0
+  firstBlockNo _ = BlockNo 0
+
+  validateEnvelope cfg oldTip hdr = do
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs]: Store whether we allow EBBs or not in the `BlockConfig` and use that here for a stricter check.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs [hunk 2]
```diff
@@ -440,10 +481,64 @@ instance UpdateLedger TestBlock where
...
+      -- EBB shares its block number with its predecessor. The chain can also
+      -- start with a regular block, in which case it has block number 1
+      nextBlockNo :: WithOrigin (IsEBB, BlockNo) -> IsEBB -> BlockNo
+      nextBlockNo Origin             IsEBB    = BlockNo 0
+      nextBlockNo Origin             IsNotEBB = BlockNo 1
+      nextBlockNo (At (IsNotEBB, b)) IsEBB    = b
+      nextBlockNo (At (_       , b)) _        = succ b
+
+      canBeEBB :: SlotNo -> Bool
+      canBeEBB (SlotNo s) = s `mod` epochSlots == 0
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs]: Would always return `False` if EBBs are not supported (stored in the `BlockConfig`)

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs [hunk 3]
```diff
@@ -440,10 +481,64 @@ instance UpdateLedger TestBlock where
...
+      nextSlotNo :: WithOrigin (IsEBB, SlotNo) -> IsEBB -> SlotNo
+      nextSlotNo Origin          _        = SlotNo 0
+      nextSlotNo (At (IsEBB, s)) IsNotEBB = s
+      nextSlotNo (At (_    , s)) _        = succ s
+
+      -- EBB shares its block number with its predecessor. The chain can also
+      -- start with a regular block, in which case it has block number 1
+      nextBlockNo :: WithOrigin (IsEBB, BlockNo) -> IsEBB -> BlockNo
+      nextBlockNo Origin             IsEBB    = BlockNo 0
+      nextBlockNo Origin             IsNotEBB = BlockNo 1
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs]: Also must be 0. Update the generator accordingly.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs [hunk 4]
```diff
@@ -440,10 +481,64 @@ instance UpdateLedger TestBlock where
...
 
 instance HasAnnTip TestBlock where
-  -- Use defaults
+  type TipInfo TestBlock = IsEBB
+  getTipInfo = thIsEBB . unTestHeader
 
 instance ValidateEnvelope TestBlock where
-  -- Use defaults
+  -- A chain can start with an EBB with block number 0
+  firstBlockNo _ = BlockNo 0
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs]:
```suggestion
  firstBlockNo _ = error "unused"
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs [hunk 5]
```diff
@@ -781,14 +781,19 @@ between :: forall blk. HasHeader blk
         -> Either (UnknownRange blk) [blk]
 between (SecurityParam k) from to m = do
     fork <- errFork
-    if Fragment.forksAtMostKBlocks k currentFrag fork
+    if partOfCurrentChain fork || Fragment.forksAtMostKBlocks k currentFrag fork
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs]: Mention #871.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs [hunk 6]
```diff
@@ -954,12 +954,17 @@ precondition Model {..} (At cmd) =
...
-        Right blks -> forall blks $ \blk -> Boolean $
-          Map.notMember (blockHash blk) $
-          forgetFingerprint (Model.invalid dbModel)
+        case Model.between secParam from to' dbModel of
+          Left  _    -> Bot
+          -- All blocks must be valid
+          Right blks -> forall blks $ \blk -> Boolean $
+            Map.notMember (blockHash blk) $
+            forgetFingerprint (Model.invalid dbModel)
+      where
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs]: Mention #871.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Reopen.hs [hunk 7]
```diff
@@ -125,6 +125,10 @@ reopen (CDBHandle varState) launchBgTasks = do
         replayed <- LgrDB.reopen cdbLgrDB cdbImmDB lgrReplayTracer
         traceWith cdbTracer $ TraceOpenEvent OpenedLgrDB
 
+        -- Drop scheduled chain selections, because we're not guaranteed that
+        -- these blocks still exist in the VolatileDB.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Reopen.hs]: Mention #1880.

**Comment by mrBliss** [comment]
I have added two "Feedback: .." commits to make reviewing easier. When approved, I will squash each in the commit right before it.
